Update Private Network Access article (#1047)
GoogleChrome/developer.chrome.com: site/en/blog/private-network-access-update/index.md

---
layout: 'layouts/blog-post.njk'

> title: "Private Network Access update: Introducing a deprecation trial"

Reviewer: ```suggestion
title: "Private Network Access update: Introducing Deprecation Trial"
```

authors:
  - titouan
description: Chrome is deprecating access to private network endpoints from non-secure websites in Chrome 93 as part of the Private Network Access specification. Read on for recommended actions.

> date: 2021-03-11

Reviewer: ```suggestion
date: 2021-07-29
```

Reviewer: I think it's better to bump this date, so the post is resurfaced on /blog

Author: This sounds good to me.

> description: Chrome is deprecating access to private network endpoints from non-secure public websites in Chrome 94 as part of the Private Network Access specification. Read on for recommended actions.

Reviewer 2: add:

```
tags:
- chrome-94
```

So it'll show up on the right tagged page.

Reviewer 3: +1.

> updated: 2021-07-29

Reviewer 3: ```suggestion
updated: 2021-08-25
```

hero: image/YLflGBAPWecgtKJLqCJHSzHqe2J2/dwtN0NkxkBmIz1EyhzAm.jpg
alt: A private sign in German
---

**Updates**

> - **July 29, 2021**: Updated timeline is announced. This article has had a major
>   rewrite.

Reviewer: ```suggestion
- **July 29, 2021**: Updated timeline announcement and introduction of a deprecation trial.
```

> - **July 29, 2021**: Updated timeline announcement and introduction of Deprecation Trial.

Reviewer 3: ```suggestion
- **August 25, 2021**: Updated timeline announcement and introduction of Deprecation Trial.
```

  blocking private network requests initiated from non-secure contexts has been changed.
  Such requests will be blocked in Chrome 93 instead of 92.

Chrome is deprecating access to private network endpoints from non-secure
websites as part of the [Private Network
Access](https://wicg.github.io/private-network-access/) specification. The aim
is to protect users from cross-site request forgery (CSRF) attacks targeting
routers and other devices on private networks. These [attacks have affected
hundreds of thousands of
users](https://securityaffairs.co/wordpress/22743/cyber-crime/soho-pharming-attack.html),
allowing attackers to redirect them to malicious servers.

Chrome will introduce the following changes:

* Blocking requests to private networks from insecure websites starting in
  Chrome 93.
* Introducing a [Deprecation Trial](#whats-deprecation-trial) which will end in Chrome

>   99. It will allow developers to request a time extension for chosen origins,

Reviewer 3: I pinged the thread for an answer. We can still set this to `100` in the meantime.

  99. It will allow developers to request a time extension for chosen origins,
  which will not be affected during the Deprecation Trial.
* Introducing a Chrome enterprise policy which will allow managed Chrome
  deployments to bypass the deprecation permanently. Available in Chrome 92.

If you need more time to mitigate the impact of the deprecation register for the
Deprecation Trial.

> If you have [a valid reason] re-enable the feature using Chrome enterprise policy.

Reviewer: ```suggestion
If you have administrative control over your users via Chrome Enterprise,
you can re-enable the feature using Chrome enterprise policy.
```

To mitigate the impact of the new restrictions, use one of the following
strategies:

* [Upgrade your website to HTTPS, and if necessary the target
  server](#both-ends-https).
* [Upgrade your website to HTTPS and use WebTransport](#web-transport).
* [Reverse the embedding relationship](#reverse-embedding).

## Timeline

* November 2020: [Call for feedback](https://web.dev/cors-rfc1918-feedback/)
  about the upcoming changes.
* March 2021: After reviewing feedback and doing outreach, upcoming changes are
  announced. The specification is renamed from CORS-RFC1918 to Private Network
  Access.
* April 2021: Chrome 90 rolls out to Stable, surfacing deprecation warnings.
* June 2021: Chrome 92 rolls out to Beta, forbidding private network requests
  from insecure contexts. After feedback from developers requesting more time to
  adjust, the deprecation is deferred to Chrome 93, to be accompanied with a
  Deprecation Trial.
* July 2021: Chrome 93 rolls out to Beta. Web developers can start signing up
  for the Deprecation Trial.
* August 2021: Chrome 93 rolls out to Stable. Web developers should have signed

>   up for the Deprecation Trial and deployed trial tokens to production.

Reviewer 3: This suggestion still stands.

* February 2022: Chrome 99 rolls out to Beta.
* March 2022: Chrome 99 rolls out to Stable. The Deprecation Trial ends. Chrome

>   blocks all private network requests from non-secure contexts.

Reviewer 3: As mentioned above, I pinged the blink-dev thread.

websites by default. Developers who still need to use the affected features must
sign up for the Deprecation Trial and obtain tokens for specified web origins,
then modify their websites to serve those tokens in HTTP headers or meta tags
(except in this case). If a website serves valid tokens matching their origin,
Chrome will allow the use of the deprecated feature for a limited amount of
time.

For more information, check out [Getting started with Chrome's origin
trials](https://developer.chrome.com/blog/origin-trials/) and the [web developer
guide to Origin
Trials](https://github.com/GoogleChrome/OriginTrials/blob/gh-pages/developer-guide.md)
for instructions.

## What's changing in Chrome

### Chrome 93

Starting in Chrome 93, non-secure contexts (broadly, websites that are not
delivered over HTTPS) are forbidden from making requests to the private network.
This was previously planned for Chrome 92, hence deprecation messages might
still mention the earlier milestone.

This deprecation is accompanied by a Deprecation Trial, allowing web developers
whose websites make use of the deprecated feature to continue using it until

> Chrome 99 by registering for tokens. See [below](#register-deprecation-trial)

Reviewer 3: Still pending.

for instructions on how to register and enable the trial on your website.

A pair of [Chrome enterprise policies](#enterprise-policies) can be leveraged to
disable the deprecation either entirely or on specific origins, indefinitely.
This allows managed Chrome installations, for example, those in corporate
settings, to avoid breakage.

> ### Chrome 99

Reviewer 3: Same as above.

a corresponding translation layer can convert the WebTransport messages to HTTP
requests.

We acknowledge that this represents a fair amount of work, but it should be
significantly easier than building on top of WebRTC; our hope is also that some
amount of the necessary investment gets implemented as reusable libraries. We
also believe it especially worthwhile considering the fact that non-secure
contexts are likely to lose access to more and more web platform features as the
platform moves toward encouraging HTTPS use in stronger ways over time.
Regardless of Private Network Access, this would likely be a wise investment
anyway.

We expect WebTransport over HTTP/3 to ship in Chrome 96 (it has begun an [Origin
Trial](https://groups.google.com/a/chromium.org/g/blink-dev/c/aaLFxzw5zL4/m/H3V_l-qlAgAJ))
with mitigations to protect against key sharing and other substandard security
practices, including:

* A short maximum expiration time for pinned certificates.
* A browser-specific mechanism for revoking certain keys that have been subject
  to abuse.

We will not ship the secure context restriction until at least two milestones

> after WebTransport is fully rolled out. This is why the Deprecation Trial is
> currently planned to expire in Chrome 99.

Reviewer 3: ```suggestion
after WebTransport is fully rolled out. The Deprecation Trial will be extended if
need be.
```

This solution does not require any administrative control over the network, and
can be used when the target server is not powerful enough to run HTTPS.

Instead of fetching private subresources from a public web app, a skeleton of
the app can be served from the private server, which then fetches all its
subresources (such as scripts or images) from a public server, such as a CDN.
The resulting web app can then make requests to the private server, as these are
considered same-origin. It can even make requests to other servers with private
IPs (but not localhost), though this might change in the long term.

By hosting only a skeleton on the private server, you can update the web app by
pushing new resources to the public server, just as you would update a public
web app. On the other hand, the resulting web app is not a secure context, so it
doesn't have access to some of the more powerful features of the web.

## Plans for the future

Restricting private network requests to secure contexts is only the first step in launching Private Network Access. Chrome is working towards implementing the rest of the specification in the coming months. Stay tuned for updates!

> 

Reviewer 3: ```suggestion

### Restricting localhost access from private websites

The changes in Chrome 94 only affect *public* websites accessing private IP addresses
or localhost. The Private Network Access specification also classifies requests from
*private* websites to localhost as problematic. Chrome will eventually deprecate these
too. This presents a slightly different set of challenges however, as many private
websites do not have domain names, complicating the use of Deprecation Trial tokens.

```

Reviewer 3: We should still add this paragraph.
